:sparkles: 视频评论，视频评论点赞/点踩功能

cfdxkk opened the pull request.

cfdxkk committed:
:sparkles: 增加视频评论相关功能的接口

cfdxkk committed:
:sparkles: 视频评论，视频评论点赞/踩

otomad commented on components/Creation/CreationComments/CreationCommentsItem.vue:
```diff
@@ -46,15 +52,25 @@
-		// TODO: We should separate upvote and downvote value!
+		if (button === "upvote") {
+			const notUpvoted = !isUpvoted.value; // TODO HaHa, 仅用于消除 ESLint 对单行分支语句的报错
+			if (notUpvoted && userSelfInfoStore.isLogined) { // 未登录，或者已经点过赞过不允许再点赞 // TODO 需要取消点赞吗？
```
需要。

otomad commented on components/Creation/CreationComments/CreationCommentsItem.vue:
```diff
@@ -46,15 +52,25 @@
+			}
+		} else {
+			const notDownvoted = !isDownvoted.value; // TODO HaHa, 仅用于消除 ESLint 对单行分支语句的报错
+			if (notDownvoted && userSelfInfoStore.isLogined) { // 未登录，或者已经点过踩过不允许再点踩 // TODO 需要取消点踩吗？
```
需要。

otomad commented:
@cfdxkk 

注意一下：当评论已经处于「加分」状态时，如果用户点击「减分」按钮，此时会取消原有的加分状态并再减一分，相当于一次性减两分。同理当评论已经处于「减分」状态时，点击「加分」按钮，会取消原有的减分状态并再加一分，相当于一次性加两分。

当然，已经处于「加分」或「减分」状态时再次点击对应按钮，理应「取消加分」「取消减分」。

otomad reviewed (changes requested):
@cfdxkk 

注意一下：当评论已经处于「加分」状态时，如果用户点击「减分」按钮，此时会取消原有的加分状态并再减一分，相当于一次性减两分。同理当评论已经处于「减分」状态时，点击「加分」按钮，会取消原有的减分状态并再加一分，相当于一次性加两分。

当然，已经处于「加分」或「减分」状态时再次点击对应按钮，理应「取消加分」「取消减分」。

如果以上功能确定已经实现了我就合并了。

cfdxkk committed:
:fire: 修改点赞/点踩为非阻塞

cfdxkk committed:
:sparkles: 发送视频评论功能

cfdxkk committed:
:bug: 修复了一个评论数量为 0 时，分页页数也为 0, 跳转进入报错页面的 bug

cfdxkk committed:
:sparkles: ①评论不再在服务端获取 ②增加一个发送评论的事件总线，发送的评论会即时显示 ③退二进一

:sparkles:
①评论不再在服务端获取
②增加一个发送评论的事件总线，发送的评论会即时显示
③退二进一（由点赞/点踩改为其反向操作时，除了本身 +1 之外，其反操作会 -1）

cfdxkk commented:
> @cfdxkk
> 
> 注意一下：当评论已经处于「加分」状态时，如果用户点击「减分」按钮，此时会取消原有的加分状态并再减一分，相当于一次性减两分。同理当评论已经处于「减分」状态时，点击「加分」按钮，会取消原有的减分状态并再加一分，相当于一次性加两分。
> 
> 当然，已经处于「加分」或「减分」状态时再次点击对应按钮，理应「不加分」「不减分」。
> 
> 如果以上功能确定已经实现了我就合并了。

@otomad 
已改为 “退二进一”
当只是执行 点赞/点踩 操作时，该操作的计数器 +1
当执行 点赞/点踩 操作之前，已经执行了其反操作，则该操作的计数器 +1，该操作的反操作的计数器 -1

点赞/点踩 记录必须与后端同步，但前端请求不阻塞，为了防止用户重复提交，现在添加了悲观锁。
用户连续对同一条评论执行 点赞/点踩 操作时，必须要等到上一个 点赞/点踩  请求完成。
触发悲观锁会有报错提示：“操作过于频繁，请稍后再试”

![GIF 2023-12-24 23-45-37](https://github.com/KIRAKIRA-DOUGA/KIRAKIRA-Cerasus/assets/41364445/effa3be9-6be9-4c96-92de-8c68e4eb738f)

cfdxkk committed:
:fire: delete console.log

otomad commented:
再次点击用户已点击的操作，是否可取消该次点击了？

otomad commented:
> “操作过于频繁，请稍后再试”

血压上来了（笑）。

能否实现对频繁的操作放于队列中，等上一个操作完成后立即发送？
或者取消上一个操作事件，然后把两个操作合并为一个操作一起重新发送？（我想这一条应该不可能实现）。

cfdxkk commented:
> 再次点击用户已点击的操作，是否可取消该次点击了？

没做，后端程序写完了，但没暴露接口。
今晚下班继续搞

cfdxkk commented:
> > “操作过于频繁，请稍后再试”
> 
> 血压上来了（笑）。
> 
> 能否实现对频繁的操作放于队列中，等上一个操作完成后立即发送？ 或者取消上一个操作事件，然后把两个操作合并为一个操作一起重新发送？（我想这一条应该不可能实现）。


① 请求队列？
理论上可以，但是可能需要轮询？会有额外性能开销。
② 幂等计算后节流发送？
理论上来说可以做到，算法复杂，不想写

不想继续研究，我觉得先暂且这样吧，以后有时间研究一下②

cfdxkk committed:
:sparkles: 视频评论 取消点赞 / 取消点踩 功能和其他问题修复

cfdxkk commented:
> > 再次点击用户已点击的操作，是否可取消该次点击了？
> 
> 没做，后端程序写完了，但没暴露接口。 今晚下班继续搞

@otomad 
第二次点击点赞/点踩按钮会取消 点赞/点踩 操作

@Aira-Sakuranomiya 
请求修复 #114 
否则难以判断 点赞/点踩 是否处于激活状态。
